src/main/java/org/sagebionetworks/bridge/services/TemplateService.java

```diff
@@ -0,0 +1,254 @@
+            throw new EntityNotFoundException(Template.class);
+        }
+        checkNotDeletingDefault(template, studyId);
+        
+        template.setStudyId(studyId.getIdentifier());
+        template.setModifiedOn(getTimestamp());
+        // no reason for these to be updated after creation
+        template.setTemplateType(existing.getTemplateType());
+        template.setCreatedOn(existing.getCreatedOn());
+        
+        Study study = studyService.getStudy(studyId);
+        Set<String> substudyIds = substudyService.getSubstudyIds(study.getStudyIdentifier());
+        
+        TemplateValidator validator = new TemplateValidator(study.getDataGroups(), substudyIds);
+        Validate.entityThrowingException(validator, template);
+        
+        persistCriteria(template);
+        templateDao.updateTemplate(template);
+        
+        return new GuidVersionHolder(template.getGuid(), new Long(template.getVersion()));
+    }
+    
+    public void deleteTemplate(StudyIdentifier studyId, String guid) {
+        // This not only verifies the template exists, it verifies it is in the caller's study
+        Template existing = getTemplate(studyId, guid);
+        if (existing.isDeleted()) {
+            throw new EntityNotFoundException(Template.class);
+        }
+        checkNotDeletingDefault(existing, studyId);
+        
+        existing.setDeleted(true);
+        existing.setModifiedOn(getTimestamp());
+        
+        templateDao.updateTemplate(existing);
+    }
+    
+    public void deleteTemplatePermanently(StudyIdentifier studyId, String guid) {
+        // Deletes fail quietly, we don't throw 404s if they're unnecessary 
+        Optional<Template> optional = templateDao.getTemplate(studyId, guid);
+        if (!optional.isPresent()) {
+            return;
+        }
+        // You cannot delete the default template (logical or physical).
+        Template existing = optional.get();
+        checkNotDeletingDefault(existing, studyId);
+        templateDao.deleteTemplatePermanently(studyId, guid);
+    }
+
+    private void checkNotDeletingDefault(Template template, StudyIdentifier studyId) {
+        Study study = studyService.getStudy(studyId);
+        String defaultGuid = study.getDefaultTemplates().get(template.getTemplateType().name().toLowerCase());
+        if (template.getGuid().equals(defaultGuid)) {
+            throw new ConstraintViolationException.Builder().withMessage("The default template for a type cannot be deleted.")
+                .withEntityKey("guid", template.getGuid()).build();
+        }
+    }
+    
+    private String getKey(Template template) {
+        return "template:" + template.getGuid();
+    }
+    
+    private Criteria persistCriteria(Template template) {
+        Criteria criteria = template.getCriteria();
+        if (criteria == null) {
+            criteria = Criteria.create();
+        }
+        criteria.setKey(getKey(template));
```

Comment:
Unit tests for create and update should check that we set the criteria key.

Reply:
added